File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    # 현재 다리를 건너고 있는 트럭들을 나타내는 큐 (트럭 무게 + 다리 위에 있는 시간)
    bridge = deque([0] * bridge_length)
    cur_weight = 0  # 현재 다리 위에 있는 트럭의 총 무게
    time = 0  # 경과 시간

    for truck in truck_weights:
        while True:
            # 1초 경과, 다리 위에 있는 가장 앞 트럭이 다리를 빠져나옴
            time += 1
            cur_weight -= bridge.popleft()

            # 현재 트럭이 다리 위에 오를 수 있는지 확인
            if cur_weight + truck <= weight:
                bridge.append(truck)
                cur_weight += truck
                break
            else:
                # 무게를 초과하면 트럭이 다리에 오르지 않고 다리에서 0만 추가
                bridge.append(0)

    # 마지막 트럭이 다리를 완전히 건너기까지의 시간 추가
    time += bridge_length

    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
// ** 다리에 완전히 오르지 않은 트럭의 무게는 무시
// bridge_length : 다리에 올라갈 수 있는 트럭의 최대 수
// weight : 견딜 수 있는 무게(weight 이하까지)
// truck_weights : 트럭 별 무게
// return : 모든 트럭이 다리를 건너려면 최소 몇 초가 걸리는지 return
class Solution {
    class Node {
        final int weight;
        final int time;
        public Node(int weight, int time) {
            this.weight = weight; this.time = time;
        }
    }
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        ArrayDeque<Integer> q = new ArrayDeque<>();
        int total = truck_weights[0];
        int time = 1;
        for(int i = 0; i < bridge_length - 1; i++) q.addLast(0);
        q.addLast(truck_weights[0]);
        int idx = 1;
        while(!q.isEmpty()) {
            time++;
            total -= q.pollFirst();
            if(idx < truck_weights.length) {
                if(total + truck_weights[idx] <= weight) {
                    total += truck_weights[idx];
                    q.addLast(truck_weights[idx]);
                    idx++;
                } else {
                    q.addLast(0);
                }
            }
        }
        return time;
    }
}